File: easy/ts/diameter_of_binary_tree.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}


function diameterOfBinaryTree(root: TreeNode | null): number {
  let max = 0;
  function dfs(root: TreeNode | null): number {
    if (!root) return 0;

    const left: number = dfs(root.left);
    const right: number = dfs(root.right);

    max = Math.max((left + right), max)

    return Math.max(left, right) + 1;
  };

  dfs(root);
  return max;
};

File: easy/ruby/diameter_of_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def diameter_of_binary_tree(root)
  @diameter = 0

  helper = lambda do |node|
    return 0 if node.nil?

    left = helper.call(node.left)
    right = helper.call(node.right)

    @diameter = [@diameter, left + right].max

    [left, right].max + 1
  end

  helper.call(root)
  @diameter
end
